### ISSUE TITLE
Remove unnecessary grouping in SVG sources

### ISSUE THREAD

**Issue by acabal** [drgrigg/rudyard-kipling_just-so-stories#9]
In some (all?) of the SVGs, I see that there's path elements but surrounded by one group element with a transform. Can you apply the transform to the children, and remove the `<g>` element?  I think you can do that either with Inkscape in a menu somewhere, or maybe with svgo: https://github.com/svg/svgo

Additionally we prefer height/width of SVGs to be whole numbers so that future edits/calculations are easier to do. This is a little more time-consuming to adjust but you  can do that in Inkscape too, because you'll have to preserve the aspect ratio. You'd open the image in Inkscape, select all, check the 'keep aspect ratio' button, and then adjust either the height or width to be a whole number. The other number will remain a decimal, but then you can uncheck the 'keep aspect ratio' button and tweak it to be a whole number too. Technically doing so will distort the aspect ratio slightly, but almost always this change is just a thousandth of a fraction and has no noticeable effect on the final rendering. If the distortion is too much then we can leave the one height/width as a fraction.

Once we do all that we can also probably remove the `preserveAspectRatio`, `height`, and `width` attributes on the SVG. For an example of the kind of code style we're looking for, see [Arthur Gordon Pym](https://raw.githubusercontent.com/standardebooks/edgar-allan-poe_the-narrative-of-arthur-gordon-pym-of-nantucket/master/src/epub/images/heiroglyphic-1.svg)

That's going to be a lot of pretty tedious work because there's a lot of images, so if you don't have time let me know and I can do that for you.

**Comment by acabal** [comment]
BTW svgo has a web app that might be easier: https://jakearchibald.github.io/svgomg/

**Comment by drgrigg** [comment]
Hmmm... For whatever reason I can't get Inkscape to run properly in GUI mode on my Mac. It works fine, of course, as a command line tool driven by build-images, but for anything else it's unusable.

So I've been using a GUI interface to Potrace to convert images to vectors, both for Just So Stories and now for Enchanted Castle, as it's possible to tweak settings in the GUI and see an immediate visual result. If you're telling me that it generates incorrect or highly inefficient SVG files, then I'm not sure what I'm going to do in future, other than avoid producing works which include illustrations.

Anyway, I'll try the utilities you've pointed me to and see if that solves the problem.

Regards,

David

**Comment by acabal** [comment]
Well they're not "incorrect" per se, but they do have more than a little cruft that we would like to avoid in the interest of clean and minimal markup.

Inkscape is really a great tool, fairly complex but worth trying to get installed. In Inkscape you could remove the groups with a single command. Just select all and do an "ungroup", which will apply the group's transforms to its children and remove the grouping. Do note that when saving from within Inkscape, select "plain svg" as the output, otherwise Inkscape adds a massive amount of its own junk markup to the SVG.

**Comment by acabal** [comment]
Also if you just can't get Inkscape working then let me know and I can put cleaning these SVGs on my own to do list.

**Comment by drgrigg** [comment]
I’m perfectly happy to look at them, I’m just not sure what I would need to do.

I ran one image from JSS through the online svgo tool (which looks nice), and after needing to change the settings to preserve the <title> tag, I got the result attached. Notably, it’s abut 30% larger in file size than the original. But does it meet your requirements? If so, I’ll just run everything through the online tool, both for JSS and for Enchanted Castle.

Regards,

David

**Comment by acabal** [comment]
I don't think you can send attachments via Github, but you can open up the SVG source and see for yourself. Is there still a `<g>` element? Are there height and width attributes on the root `svg` element?

**Comment by drgrigg** [comment]
This is what the top of it looks like in a text editor. Looks fine to me (but what do I know?)

<svg xmlns="http://www.w3.org/2000/svg" height="808.049" version="1" width="1130.314">
  <title>
    The Whale swallowing the Mariner with his infinite-resource-and-sagacity.
  </title>
  <path d="M355.762 4.16c-.953 2.38-1.715 4.57-1.715 5.047 0 .38-.952 1.619-2.095 2.762-1.143 1.143-1.905 2.762-1.524 3.524.286.762.095 1.428-.476 1.428-.476 0-.952.762-.952 1.714 0 .858-.667 2.381-1.429 3.239-.762.857-1.714 2.857-2.19 4.571l-.667 2.952 2.095-1.904c1.238-1.048 2.19-1.524 2.19-1.048 0 .38-1.142 1.714-2.57 2.952-1.715 1.429-2.763 3.62-3.144 6.381-.285 2.286-.857 4.096-1.333 4.096-.857 0-4.38 5.714-5.143 8.571-.286 1.048-2.666 8.762-5.333 16.952-4.762 15.238-7.81 27.524-6.953 28.381.286.286.572.286.762.095.19-.285 2-2.57 4-5.142 4-5.048 5.048-8.381 1.715-5.334-3.429 3.143-2.19.572 2-4.19 2.19-2.476 4-5.048

Regards,

David

**Comment by acabal** [comment]
Looks pretty good though Github cut it off halfway. We would want to replace the height/width attributes with a single viewbox attribute though. That can be scriptable so you don't have to do it by hand for each one.

**Comment by drgrigg** [comment]
Gordon Pym uses a ViewBox, I see. Should I turn on that setting on svgo? And should I then remove the precision after the decimal point for width and height?

Regards,

David

**Comment by drgrigg** [comment]
Sorry, we cross-posted!  I only included the first few lines of the file, it hadn’t been cut off. I’ll run the tool again with ViewBox turned on and then send you the whole file.

Regards,

David

**Comment by acabal** [comment]
Give it a shot. I don't know if simply removing the precision is sufficient as that may distort the aspect ratio and make the image look weird. You'd have to eyeball it.

**Comment by drgrigg** [comment]
We’re only talking about fractions of one percent, I’m sure you’d never notice. I’ve edited this one manually and the result looks fine to me. See attached file. If you’re happy with this, I’ll run all the images through the tool and write a script to remove the precision in the view box.

Regards,

David

**Comment by acabal** [comment]
OK, go for it!
